ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,203 @@
+    show (FetchingNewLedgerState cnt) =
+        printf "Fetching new ledgerstate, %d registered pools"
+            cnt
+
+
+data RelayAddress = RelayAddressDomain DomainAddress
+                  | RelayAddressAddr IP.IP Socket.PortNumber
+                  deriving (Show, Eq, Ord)
+
+-- | The relative stake of the stakepool. A value in the [0, 1] range.
+newtype PoolStake = PoolStake { unPoolStake :: Rational } deriving (Eq, Fractional, Num, Ord, Show)
+
+-- | The relative stake of the stakepool and all preceeding pools. A value in the range [0, 1].
+newtype AckPoolStake = AckPoolStake { unAckPoolStake :: Rational } deriving (Eq, Num, Ord)
+
+-- | Convert a list of pools with stake to a Map keyed on the ackumulated stake.
+--
+ackPoolStake :: [(PoolStake, NonEmpty RelayAddress)]
+             -> Map AckPoolStake (PoolStake, NonEmpty RelayAddress)
+ackPoolStake pl =
+    let pl' = reRelativeStake pl
+        ackList = snd $ foldl' fn (AckPoolStake 0, []) pl' in
+    Map.fromList ackList
+  where
+    fn :: (AckPoolStake, [(AckPoolStake, (PoolStake, NonEmpty RelayAddress))])
+       -> (PoolStake, NonEmpty RelayAddress)
+       -> (AckPoolStake, [(AckPoolStake, (PoolStake, NonEmpty RelayAddress))])
+    fn (as, ps) (s, rs) =
+        let ackst = AckPoolStake (unPoolStake s) in
+        (as + ackst, (as + ackst, (s, rs)) : ps)
+
+-- | Not all stake pools have valid/usable relay information. This means that we need to
+-- recalculate the relative stake for each pool.
+reRelativeStake :: [(PoolStake, NonEmpty RelayAddress)]
+                -> [(PoolStake, NonEmpty RelayAddress)]
+reRelativeStake pl =
+    let total = sum $ map fst pl in
+    map (\(s, rls) -> (s / total, rls)) pl
+
+-- try to pick n random peers
+pickPeers :: forall m. Monad m
+          => StdGen
+          -> Tracer m TraceLedgerPeers
+          -> Map AckPoolStake (PoolStake, NonEmpty RelayAddress)
+          -> Word16
+          -> m (StdGen, [RelayAddress])
+pickPeers inRng _ pools _ | Map.null pools = return (inRng, [])
+pickPeers inRng tracer pools cnt = go inRng cnt []
+  where
+    go :: StdGen -> Word16 -> [RelayAddress] -> m (StdGen, [RelayAddress])
+    go rng 0 picked = return (rng, picked)
+    go rng n picked =
+        let (r :: Word64, rng') = random rng
+            d = maxBound :: Word64
+            x = fromIntegral r % fromIntegral d in
+        case Map.lookupGE (AckPoolStake x) pools of
+             Nothing -> go rng' (n - 1) picked -- XXX We failed pick a peer. Shouldn't this be an error?
+             Just (ackStake, (stake, relays)) -> do
+                 let (ix, rng'') = randomR (0, NonEmpty.length relays - 1) rng'
+                     relay = relays NonEmpty.!! ix
+                 traceWith tracer $ PickedPeer relay ackStake stake
+                 go rng'' (n - 1) (relay : picked)
+
+-- | Example of how how peers from ledger could be used.
+-- In reality it will be implemented using the new connection-manager and governor.
```

Comment:
```suggestion
-- | Example of how how peers from ledger could be used.
-- In reality it will be implemented using the new connection-manager and governor.
--
```